src/SignalR/common/Http.Connections.Common/src/NegotiateProtocol.cs

```
@@ -13,77 +15,166 @@ namespace Microsoft.AspNetCore.Http.Connections
+                                Skip(ref reader);
                             }
-
-                            jsonWriter.WriteEndArray();
-                            jsonWriter.WriteEndObject();
-                        }
+                            break;
+                        case JsonTokenType.EndObject:
+                            completed = true;
+                            break;
+                        default:
+                            throw new InvalidDataException($"Unexpected token '{reader.TokenType}' when reading negotiation response JSON.");
                     }
+                }
 
-                    jsonWriter.WriteEndArray();
-                    jsonWriter.WriteEndObject();
+                if (url == null && error == null)
+                {
+                    // if url isn't specified or there isn't an error, connectionId and available transports are required
+                    if (connectionId == null)
+                    {
+                        throw new InvalidDataException($"Missing required property '{ConnectionIdPropertyName}'.");
+                    }
 
-                    jsonWriter.Flush();
+                    if (availableTransports == null)
+                    {
+                        throw new InvalidDataException($"Missing required property '{AvailableTransportsPropertyName}'.");
+                    }
                 }
+
+                return new NegotiationResponse
+                {
+                    ConnectionId = connectionId,
+                    Url = url,
+                    AccessToken = accessToken,
+                    AvailableTransports = availableTransports,
+                    Error = error,
+                };
             }
-            finally
+            catch (Exception ex)
             {
-                Utf8BufferTextWriter.Return(textWriter);
+                throw new InvalidDataException("Invalid negotiation response received.", ex);
```

Why do we need this catch all exception block now when it wasn't needed previously?

---

It was there before, but the git diff makes it hard to see.